AnalyzerTools/src/MCCorrection.C

```diff
@@ -843,3 +852,321 @@ double MCCorrection::GetOfficialDYReweight(const vector<Gen>& gens, int sys){
+    }
+  }
+
+  return Prob_DATA/Prob_MC;
+
+}
+
+bool MCCorrection::IsBTagged_2a(JetTagging::Parameters jtp, const Jet& jet, string Syst){
+
+  double this_discr = jet.GetTaggerResult(jtp.j_Tagger);
+  double cutValue = GetJetTaggingCutValue(jtp.j_Tagger, jtp.j_WP);
+
+  bool isBTagged = this_discr > cutValue;
+
+  if(IsDATA) return isBTagged;
+
+  //==== Set seed
+  unsigned int runNum_uint  = static_cast <unsigned int> (run);
+  unsigned int lumiNum_uint = static_cast <unsigned int> (lumi);
+  unsigned int evNum_uint   = static_cast <unsigned int> (event);
+  unsigned int jet0eta = uint32_t(fabs(jet.Eta())/0.01);
+  int m_nomVar=1;
+  std::uint32_t seed = jet0eta + m_nomVar + (lumiNum_uint<<10) + (runNum_uint<<20) + evNum_uint;
+
+  TRandom3 rand_(seed);
+
+  bool newBTag = isBTagged;
+
+  //=== Get SF
+  double Btag_SF =  GetJetTaggingSF(jtp,
+                                    jet.hadronFlavour(),
+                                    jet.Pt(),
+                                    jet.Eta(),
+                                    jet.GetTaggerResult(jtp.j_Tagger),
+                                    Syst );
+
+
+  if(Btag_SF == 1) return newBTag; //no correction needed
+
+  //=== throw random number to apply correction
+  float coin = rand_.Uniform(1.);
+  if(Btag_SF > 1){
+    //=== use this if SF>1
+
+    if( !isBTagged ) {
+
+      double Btag_eff = GetMCJetTagEff(jtp.j_Tagger, jtp.j_WP, jet.hadronFlavour(), jet.Pt(), jet.Eta());
+      //=== fraction of jets that need to be upgraded
+      float mistagPercent = (1.0 - Btag_SF) / (1.0 - (1./Btag_eff) );
+
+      //=== upgrade to tagged
+      if( coin < mistagPercent ) {newBTag = true;}
+    }
+
+  }else{
+    //=== use this if SF<1
+ 
+    //=== downgrade tagged to untagged
+    if( isBTagged && coin > Btag_SF ) {newBTag = false;}
+
+  }
+
+  return newBTag;
+
+}
+
+double MCCorrection::GetScaledJetTaggerResult_2b(JetTagging::Parameters jtp, const Jet& jet, string Syst){
+
+  if(jtp.j_MeasurmentType_Light!=JetTagging::iterativefit || jtp.j_MeasurmentType_Heavy!=JetTagging::iterativefit){
+    cout << "[MCCorrection::GetScaledJetTaggerResult_2b] This method only works for iterativefit method" << endl;
+    exit(EXIT_FAILURE);
+    return false;
+  }
+
+  double this_discr = jet.GetTaggerResult(jtp.j_Tagger);
+  if(IsDATA) return this_discr;
+
+  double Btag_SF =  GetJetTaggingSF(jtp,
+                                    jet.hadronFlavour(),
+                                    jet.Pt(),
+                                    jet.Eta(),
+                                    jet.GetTaggerResult(jtp.j_Tagger),
+                                    Syst );
+
+  return this_discr * Btag_SF;
+
+}
+
+bool MCCorrection::IsBTagged_2b(JetTagging::Parameters jtp, const Jet& jet, string Syst){
+
+  double new_discr = GetScaledJetTaggerResult_2b(jtp, jet, Syst);
+  double cutValue = GetJetTaggingCutValue(jtp.j_Tagger, jtp.j_WP);
+
+  return new_discr > cutValue;
+
+}
+
```

Comment:
2b method is also event by event weight based method.
```
double MCCorrection::GetBTaggingReweight_2b(const vector<Jet>& jets, JetTagging::Parameters jtp, string Syst){

   if(IsDATA) return 1.;

   double weight(1.);
   for(unsigned int i=0; i<jets.size(); i++){
     
     double this_SF = GetJetTaggingSF(jtp,
                                      jets.at(i).hadronFlavour(),
                                      jets.at(i).Pt(),
                                      jets.at(i).Eta(),
                                      jets.at(i).GetTaggerResult(jtp.j_Tagger),
                                      Syst );

      weight *=this_SF;
   }

   return weight;

 }
```

Reply:
2b) is one of the " 2) Methods **without** event reweighting " isn't it?

Reply:
twiki is a bit confusing.
There are two different method in same section.
  1) One is a method **without** reweighting [presentation](https://indico.cern.ch/getFile.py/access?contribId=7&resId=0&materialId=slides&confId=162142)
  2) One is a method with reweighting [example code](https://twiki.cern.ch/twiki/bin/view/CMS/BTagShapeCalibration)

Reply:
ohh so 1d) method from [this twiki](https://twiki.cern.ch/twiki/bin/view/CMS/BTagSFMethods#1d_Event_reweighting_using_discr) is what we should do with that SFs?

Reply:
I think you are right!

Reply:
thanks, I will update this too